registry/registry.go

```diff
@@ -759,24 +836,297 @@ func downloadVersion(opts *VersionOptions) (ver *Version, attachments []*kivik.A
 		} else if tryCount <= 3 {
 			continue
 		} else {
-			return
+			return nil, err
 		}
 	}
 
+	// Reader for filesize
 	counter := &Counter{}
 	var reader io.Reader = buf
 	reader = io.TeeReader(reader, counter)
 
-	var packVersion string
+	// Reading the tarball content
+	tarball, err := ReadTarballVersion(reader, contentType, url)
+	if err != nil {
+		return nil, err
+	}
+
+	// Adding metadata to the tarball struct
+	tarball.ContentType = contentType
+	tarball.Size = counter.Written()
+
+	if !tarball.HasPrefix {
+		tarball.TarPrefix = ""
+	}
+
+	return tarball, nil
+}
+
+func downloadVersion(opts *VersionOptions) (ver *Version, attachments []*kivik.Attachment, err error) {
+	url := opts.URL
+
+	tarball, errd := downloadTarball(opts, url)
+	if errd != nil {
+		return nil, nil, errd
+	}
+
+	// Checks
+	if _, erre := tarball.CheckEditor(); erre != nil {
+		err = multierror.Append(err, erre)
+	}
+	if _, errs := tarball.CheckSlug(); errs != nil {
+		err = multierror.Append(err, errs)
+	}
+	if _, errv := tarball.CheckVersion(opts.Version); errv != nil {
+		err = multierror.Append(err, errv)
+	}
+
+	// Handling tarball assets
+	attachments, erra := HandleAssets(tarball, opts)
+	if erra != nil {
+		err = multierror.Append(err, erra)
+	}
+
+	// If there was any error during checks, we are not going further
+	if err != nil {
+		return nil, nil, err
+	}
+
+	manifestContent := tarball.ManifestContent
+	manifest := tarball.ManifestMap
+
+	// Adding custom parameters if needed
+	var errm error
+	if opts.Parameters != nil {
+		manifest["parameters"] = opts.Parameters
+		manifestContent, errm = json.Marshal(manifest)
+		if errm != nil {
+			return nil, nil, err
+		}
+	}
+
+	// Retreiving the tarball manifest
+	parsedManifest := tarball.Manifest
+
+	filename := filepath.Base(url)
+	filepath := filepath.Join(parsedManifest.Slug, parsedManifest.Version, filename)
+
+	// Saving app tarball
+	err = SaveTarball(opts.Space, filepath, tarball)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Creating version
+	ver = new(Version)
+	ver.ID = getVersionID(parsedManifest.Slug, opts.Version)
+	ver.Slug = parsedManifest.Slug
+	ver.Version = opts.Version
+	ver.Type = tarball.AppType
+	// Now the tarball has been downloaded, override the original tarball URL to
+	// local registry url for future downloads
+	ver.URL = opts.RegistryURL.String()
+	ver.Sha256 = opts.Sha256
+	ver.Editor = parsedManifest.Editor
+	ver.Manifest = manifestContent
+	ver.Size = tarball.Size
+	ver.TarPrefix = tarball.TarPrefix
+	ver.CreatedAt = time.Now().UTC()
+	return
```

Comment:
[nit] I'd prefer if we don't mix "naked" `return` (with named parameters) and `return` with explicit things